src/packagedcode/models.py

```diff
@@ -765,6 +765,251 @@ def recognize(cls, location):
         raise NotImplementedError
 
 
+@attr.s()
+class PackageInstance:
+    """
+    A package instance mixin as represented by its package manifests, files and data
+    from its package manifests.
+
+    Subclasses must extend a Package subclass for a given ecosystem.
+    """
+
+    package_uuid = String(
+        label='Package instance UUID',
+        help='A unique ID for package instances in a codebase scan.'
+             'Consists of a pURL and an UUID field as a pURL qualifier.'
+    )
+
+    package_manifest_paths = List(
+        item_type=String,
+        label='Package manifest paths',
+        help='List of package manifest file paths for this package'
+    )
+
+    files = List(
+        item_type=PackageFile,
+        label='Provided files',
+        help='List of files provided by this package.'
+    )
+
+    def get_package_data(self):
+        """
+        Returns a mapping of package data attributes and corresponding values.
+        """
+        mapping = self.to_dict()
+
+        # Removes PackageInstance specific attributes
+        for attribute in ('package_uuid', 'package_manifest_paths', 'files'):
+            mapping.pop(attribute)
+        
+        # Remove attributes which are BasePackage functions
+        for attribute in ('repository_homepage_url', 'repository_download_url', 'api_data_url'):
```

Comment:
I have no strong feelings one way or the other about this, but I am curious as to why `repository_homepage_url`, `repository_download_url`, and `api_data_url` are removed from the package data that is to be returned.

Reply:
So this is a helper function used in the merging of package_manifest data into a package_instance [here](https://github.com/nexB/scancode-toolkit/blob/2098-top-level-packages/src/packagedcode/models.py#L897). These are basically fields which are not set as values in a package data, but obtained from functions like [this](https://github.com/nexB/scancode-toolkit/blob/2098-top-level-packages/src/packagedcode/models.py#L320) in the `to_dict()`. So we don't need to merge these data from the mappings.